reactos/drivers/usb/usbohci_new/usbohci.c

```diff
@@ -0,0 +1,2752 @@
+    UCHAR ConditionCode;
+    BOOLEAN IsResetOnHalt = FALSE;
+
+    //DPRINT_OHCI("OHCI_PollAsyncEndpoint: Endpoint - %p\n", OhciEndpoint);
+
+    ED = OhciEndpoint->HcdED;
+    NextTdPA = ED->HwED.HeadPointer & OHCI_ED_HEAD_POINTER_MASK;
+
+    if (!NextTdPA)
+    {
+        OHCI_DumpHcdED(ED);
+        DbgBreakPoint();
+    }
+
+    NextTD = RegPacket.UsbPortGetMappedVirtualAddress((PVOID)NextTdPA,
+                                                      OhciExtension,
+                                                      OhciEndpoint);
+    DPRINT_OHCI("NextTD - %p\n", NextTD);
+
+    if ((ED->HwED.HeadPointer & OHCI_ED_HEAD_POINTER_HALT) == 0)
+    {
+        goto ProcessListTDs;
+    }
+
+    OHCI_DumpHcdED(ED);
+
+    IsResetOnHalt = (ED->Flags & OHCI_HCD_ED_FLAG_RESET_ON_HALT) != 0;
+    DPRINT1("PollAsyncEndpoint: IsResetOnHalt %x\n", IsResetOnHalt);
+
+    for (TD = OhciEndpoint->HcdHeadP; ; TD = TD->NextHcdTD)
+    {
+        if (!TD)
+        {
+            OHCI_DumpHcdED(ED);
+            DbgBreakPoint();
+        }
+
+        if (TD == NextTD)
+        {
+            DPRINT("TD == NextTD - %p\n", TD);
+            goto HandleDoneList;
+        }
+
+        OhciTransfer = TD->OhciTransfer;
+        ConditionCode = TD->HwTD.gTD.Control.ConditionCode;
+
+        DPRINT("TD - %p, ConditionCode - %X\n", TD, ConditionCode);
+        OHCI_DumpHcdTD(TD);
+
+        switch (ConditionCode)
+        {
+            case OHCI_TD_CONDITION_NO_ERROR:
+                TD->Flags |= OHCI_HCD_TD_FLAG_DONE;
+                InsertTailList(&OhciEndpoint->TDList, &TD->DoneLink);
+                continue;
+
+            case OHCI_TD_CONDITION_NOT_ACCESSED:
+                TD->Flags |= (OHCI_HCD_TD_FLAG_DONE | OHCI_HCD_TD_FLAG_NOT_ACCESSED);
+                InsertTailList(&OhciEndpoint->TDList, &TD->DoneLink);
+                continue;
+
+            case OHCI_TD_CONDITION_DATA_UNDERRUN:
+                DPRINT1("DATA_UNDERRUN. Transfer->Flags - %X\n", OhciTransfer->Flags);
+
+                if (OhciTransfer->Flags & OHCI_TRANSFER_FLAGS_SHORT_TRANSFER_OK)
+                {
+                    IsResetOnHalt = TRUE;
+                    TD->HwTD.gTD.Control.ConditionCode = OHCI_TD_CONDITION_NO_ERROR;
+
+                    ControlStatusTD = OhciTransfer->ControlStatusTD;
+
+                    if ((TD->Flags & OHCI_HCD_TD_FLAG_CONTROLL_STATUS) == 0 &&
+                        ControlStatusTD)
+                    {
+                        PhysicalAddress = ControlStatusTD->PhysicalAddress;
+                        PhysicalAddress |= (ED->HwED.HeadPointer &
+                                            OHCI_ED_HEAD_POINTER_FLAGS_MASK);
+
+                        ED->HwED.HeadPointer = PhysicalAddress;
+
+                        NextTD = OhciTransfer->ControlStatusTD;
+                        DPRINT("PhysicalAddress - %p, NextTD - %p\n", PhysicalAddress, NextTD);
+                    }
+                    else
+                    {
+               
```

Comment:
Implementing this sooner rather than later seems like a good idea. Loading/unloading the driver on Windows with Driver Verifier enabled is the easiest way to find leaks.

Reply:
It's not clear what - again GitHub shows another ...

Reply:
This was referring to OHCI_Unload.